presto-spi/src/main/java/com/facebook/presto/spi/block/BlockFlattener.java

```diff
@@ -0,0 +1,89 @@
+package com.facebook.presto.spi.block;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.spi.block.BlockLeases.newLease;
+
+public class BlockFlattener
+{
+    private final ArrayAllocator allocator;
+
+    public BlockFlattener(ArrayAllocator allocator)
+    {
+        this.allocator = allocator;
+    }
+
+    public BlockLease flatten(Block block)
+    {
+        if (block instanceof DictionaryBlock) {
+            return flattenDictionary((DictionaryBlock) block);
+        }
+        if (block instanceof RunLengthEncodedBlock) {
+            return flattenRLE((RunLengthEncodedBlock) block);
+        }
+        return newLease(block);
+    }
+
+    public BlockLease flattenDictionary(DictionaryBlock dictionaryBlock)
+    {
+        Block block = dictionaryBlock.getDictionary();
+        int[] map = (int[]) dictionaryBlock.getIds().getBase();
+        int positionCount = dictionaryBlock.getPositionCount();
+        int[] leasedMap = null;
+        while (true) {
+            if (block instanceof DictionaryBlock) {
+                dictionaryBlock = (DictionaryBlock) block;
+                int[] ids = (int[]) dictionaryBlock.getIds().getBase();
+
+                if (leasedMap != null) {
+                    allocator.returnArray(leasedMap);
+                }
+                int[] newMap = allocator.getIntArray(positionCount);
+                leasedMap = newMap;
+                for (int i = 0; i < positionCount; ++i) {
+                    newMap[i] = ids[map[i] + dictionaryBlock.getOffset()];
+                }
+                map = newMap;
```

Comment:
Here the leasedMap was released before calculating the newMap[i] which references to map[], but map[] and leasedMap[] points to the same array object from last level. Although returnArray() does not actually free the memory such that the code could run without throwing errors, but the semantics is better to be that release the leasedMap after the calculation is done:
```
                int[] newMap = allocator.getIntArray(positionCount);
                for (int i = 0; i < positionCount; ++i) {
                    newMap[i] = ids[map[i] + dictionaryBlock.getOffset()];
                }
                if (leasedMap != null) {
                    allocator.returnArray(leasedMap);
                }
                leasedMap = newMap;
                map = newMap;
```